### PR TITLE
[PIR]Add Fused dropout add pass

### PR DESCRIPTION
**Pull request by xiaoguoguo626807**
### PR types

Others 
### PR changes

Others 
### Description

pcard-67164
添加dropout + add ; dropout + add + add_grad + dropoutgrad 的pass

### CODE REVIEW

File: test/cpp/pir/pattern_rewrite/CMakeLists.txt [hunk 1]
```diff
@@ -24,6 +24,7 @@ cc_test_old(
   SRCS
   drr_same_type_binding_test.cc
   DEPS
+  fused_gemm_epilogue_pass
```
Comment (xiaoguoguo626807) [inline test/cpp/pir/pattern_rewrite/CMakeLists.txt:27]:
```suggestion
  fused_gemm_epilogue_pass
```

Comment (xiaoguoguo626807) [inline test/cpp/pir/pattern_rewrite/CMakeLists.txt]:
```suggestion
```

File: paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+namespace pir {
+
+std::unique_ptr<Pass> CreateFusedDropoutAddPass() {
+  return std::make_unique<FusedDropoutAddPass>();
+}
+
+}  // namespace pir
+
+REGISTER_IR_PASS(fused_dropout_add, FusedDropoutAddPass);
```
Comment (yuanlehome) [inline paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc]:
```suggestion
REGISTER_IR_PASS(fused_dropout_add_pass, FusedDropoutAddPass);
```

File: paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+  void Run(pir::Operation *op) override {
+    pir::GreedyRewriteConfig cfg;
+    cfg.use_top_down_traversal = true;
+    cfg.max_iterations = 10;
+    pir::ApplyPatternsGreedily(op->region(0), patterns_, cfg);
+  }
+
+  bool CanApplyOn(pir::Operation *op) const override {
+    return op->name() == "builtin.module" && op->num_regions() > 0;
```
Comment (yuanlehome) [inline paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc]:
```suggestion
    return op->isa<::pir::ModuleOp>() && op->num_regions() > 0;
```

File: paddle/fluid/pybind/pir.cc [hunk 4]
```diff
@@ -66,6 +67,7 @@ using pir::Value;
 using pybind11::return_value_policy;
 
 USE_PASS(dead_code_elimination_pass);
+USE_PASS(fused_dropout_add);
```
Comment (xiaoguoguo626807) [inline paddle/fluid/pybind/pir.cc]:
```suggestion
USE_PASS(fused_dropout_add_pass);
```

File: test/ir/new_ir/fused_pass/test_fused_dropout_add.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                res1 = paddle.nn.functional.dropout(x=x, p=0.5, training=True)
+                res2 = paddle.add(res1, y)
+                res3 = paddle.sum(res2)
+
+                op_names = [op.name() for op in main_program.global_block().ops]
+                self.assertTrue('pd_op.dropout' in op_names)
+                self.assertTrue('pd_op.add' in op_names)
+                pm = paddle.pir.PassManager()
+                pm.add_pass(
+                    'fused_dropout_add'
```
Comment (xiaoguoguo626807) [inline test/ir/new_ir/fused_pass/test_fused_dropout_add.py]:
```suggestion
                    'fused_dropout_add_pass'
```

File: test/ir/new_ir/fused_pass/test_fused_dropout_add.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+                self.assertTrue(
+                    'pd_op.dropout' in op_names and 'pd_op.add' in op_names
+                )
+                self.assertTrue(
+                    'pd_op.add_grad' in op_names
+                    and 'pd_op.dropout_grad' in op_names
+                )
+                pm = paddle.pir.PassManager()
+                pm.add_pass(
+                    'fused_dropout_add'
```
Comment (xiaoguoguo626807) [inline test/ir/new_ir/fused_pass/test_fused_dropout_add.py]:
```suggestion
                    'fused_dropout_add_pass'
```

File: paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+        {&res.Tensor("x"), &res.Tensor("y"), &res.Tensor("seed_tensor")},
+        {&res.Tensor("add_out"), &res.Tensor("mask")});
+    fused_dropout_add_grad({&res.Tensor("mask"), &res.Tensor("add_out_grad")},
+                           {&res.Tensor("x_grad"), &res.Tensor("y_grad")});
+  }
+};
+
+class FusedDropoutAddPass : public pir::Pass {
+ public:
+  FusedDropoutAddPass() : pir::Pass("fused_dropout_add", 1) {}
```
Comment (xiaoguoguo626807) [inline paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc]:
```suggestion
  FusedDropoutAddPass() : pir::Pass("fused_dropout_add_pass", 1) {}
```

File: paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+#include "paddle/pir/pattern_rewrite/pattern_rewrite_driver.h"
+
+namespace {
+
+class FusedDropoutAddPattern
+    : public pir::drr::DrrPatternBase<FusedDropoutAddPattern> {
+ public:
+  void operator()(pir::drr::DrrPatternContext *ctx) const override {
+    pir::drr::SourcePattern pat = ctx->SourcePattern();
+    const auto &dropout = pat.Op("pd_op.dropout",
```
Comment (Aurelius84) [inline paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc:29]:
```suggestion
    const auto &dropout = pat.Op(paddle::dialect::DropoutOp::name(),
```
其实更推荐这样来写

**Review by Aurelius84 — DISMISSED** [review]
LGTM

File: paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc (hunk 8 above)
Comment (xiaoguoguo626807) [reply paddle/fluid/pir/transforms/fusion/fused_dropout_add_pass.cc:29]: 后续可以统一修改

**Review by changeyoung98 — APPROVED** [review]

**Review by jeff41404 — APPROVED** [review]

**Review by sunzhongkai588 — APPROVED** [review]
no doc changes. LGTM
